In [1]:
# import lib
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_squared_error
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.model_selection import cross_val_score

In [2]:
data = pd.read_csv('../data/proceed/Clean_data.csv')

In [3]:
data.head()

accounting_date  company_code customer_code  customer_district_code  \
0      2012-05-09           101     411800601                     410   
1      2012-02-16           101     361000403                     300   
2      2012-05-09           101     361000403                     300   
3      2012-05-18           101     565540415                     500   
4      2012-01-09           101     565540415                     500   

     item_code       business_chain_l1_name salesperson_code  \
0  GENIE8WWWBC              ElectraCorp Ltd             T612   
1  GENIE8WWWBC        BrightPower Solutions             P345   
2  GENIE8WWWBC        BrightPower Solutions             P345   
3  GENIE8WWWBC  Global Electric Wholesalers             T504   
4  GENIE8WWWBC  Global Electric Wholesalers             T504   

            market_segment  value_sales  value_cost  ...  \
0  Commercial & Industrial       218.40    178.1976  ...   
1  Commercial & Industrial        38.28     25.4568  ...   
2  Commercial & Industrial        40.20     25.4568  ...   
3  Commercial & Industrial        20.10     12.7284  ...   
4  Commercial & Industrial        19.14     12.7284  ...   

   business_chain_l1_code_ZZ1  business_chain_l1_code_ZZ2  \
0                       False                       False   
1                       False                       False   
2                       False                       False   
3                       False                       False   
4                       False                       False   

  contact_method_code_Other  order_type_code_NOR  order_type_code_Other  \
0                     False                 True                  False   
1                     False                 True                  False   
2                     False                 True                  False   
3                     False                False                  False   
4                     False                False                  False   

    Profit Unit_Price Profit_Margin Transformed_Profit  Transformed_Unit_Price  
0  40.2024       2.60     18.407692           6.340536                1.280934  
1  12.8232       3.19     33.498433           3.580950                1.432701  
2  14.7432       3.35     36.674627           3.839687                1.470176  
3   7.3716       3.35     36.674627           2.715069                1.470176  
4   6.4116       3.19     33.498433           2.532114                1.432701  

[5 rows x 85 columns]

In [4]:
data.shape

(1971500, 85)

- Display the column name

In [6]:
data.columns

Index(['accounting_date', 'company_code', 'customer_code',
       'customer_district_code', 'item_code', 'business_chain_l1_name',
       'salesperson_code', 'market_segment', 'value_sales', 'value_cost',
       'value_quantity', 'value_price_adjustment', 'currency',
       'invoice_number', 'line_number', 'invoice_date',
       'customer_order_number', 'order_date', 'dss_update_time', 'day_gap',
       'order_group', 'value_sales_log_plus1_transformed',
       'value_cost_log_plus1_transformed',
       'value_quantity_log_plus1_transformed', 'business_area_code_DLT',
       'business_area_code_LMP', 'business_area_code_OTH',
       'business_area_code_Other', 'business_area_code_SUR',
       'item_group_code_Other', 'item_class_code_LMP01',
       'item_class_code_LMP03', 'item_class_code_LMP04',
       'item_class_code_LMP05', 'item_class_code_Other', 'item_type_5',
       'item_type_6', 'item_type_7', 'item_type_Other',
       'bonus_group_code_Trade', 'environment_group_code_Other'

- Check type of data

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1971500 entries, 0 to 1971499
Data columns (total 85 columns):
 #   Column                                  Dtype  
---  ------                                  -----  
 0   accounting_date                         object 
 1   company_code                            int64  
 2   customer_code                           object 
 3   customer_district_code                  int64  
 4   item_code                               object 
 5   business_chain_l1_name                  object 
 6   salesperson_code                        object 
 7   market_segment                          object 
 8   value_sales                             float64
 9   value_cost                              float64
 10  value_quantity                          float64
 11  value_price_adjustment                  int64  
 12  currency                                object 
 13  invoice_number                          int64  
 14  line_number                       

#### 1. Question 1: What factors drive sales revenue across different customer segments?

Objective
- The goal of this question is to understand the key drivers of sales revenue across different customer segments. By analyzing the impact of factors such as product quantity, cost, discounts, and regional segmentation, LuminaTech can identify which factors most significantly contribute to increased revenue. This insight can inform marketing strategies, inventory planning, and pricing adjustments, enabling the company to optimize revenue across diverse customer groups.

- Define X with Log-Transformed Columns

In [12]:
# Define the independent variables (features) with log-transformed columns
X = data[['value_quantity_log_plus1_transformed', 'value_cost_log_plus1_transformed', 
                  'value_price_adjustment', 'day_gap', 
                  'business_area_code_DLT', 'business_area_code_LMP', 
                  'business_area_code_OTH', 'business_area_code_Other', 
                  'business_area_code_SUR']]


y = data['value_sales_log_plus1_transformed']

- Split Data into Training and Testing Sets

In [14]:
# Perform an 80-20 split for training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

- Build and Fit the Multiple Regression Model

In [16]:
# Initialize the linear regression model
model = LinearRegression()

# Fit the model on the training data
model.fit(X_train, y_train)

LinearRegression()

- Evaluate the Model’s Performance

In [18]:
# Make predictions on the test set
y_pred = model.predict(X_test)

# Calculate R-squared and RMSE (Root Mean Squared Error) for performance evaluation
r2 = r2_score(y_test, y_pred)
rmse = mean_squared_error(y_test, y_pred, squared=False)

print(f"R-squared: {r2}")
print(f"RMSE: {rmse}")

R-squared: 0.2805906505343214
RMSE: 415.45364863705595


R-squared & RMSE
- R-squared (0.281): This indicates that approximately 28.1% of the variance in value_sales_log_plus1_transformed (log-transformed sales revenue) is explained by the selected features. While this isn't very high, it suggests that other factors not included in the model might also influence sales revenue.

In practical terms, the model explains a limited portion of the revenue drivers, implying the need for additional variables or more complex models if higher predictive power is desired.
- RMSE (415.45): The Root Mean Squared Error (RMSE) provides a measure of the model’s average prediction error. A lower RMSE would indicate better model fit. Here, an RMSE of 415.45 suggests that there is a moderate level of error in the model’s predictions of sales revenue.

- Interpret the Coefficients

In [21]:
# Print the coefficients along with their respective feature names
coefficients = pd.DataFrame({"Feature": X.columns, "Coefficient": model.coef_})
print(coefficients)

                                Feature  Coefficient
0  value_quantity_log_plus1_transformed    -0.596836
1      value_cost_log_plus1_transformed     1.603947
2                value_price_adjustment  -180.585334
3                               day_gap     0.034158
4                business_area_code_DLT    -1.611231
5                business_area_code_LMP     4.287038
6                business_area_code_OTH     0.047396
7              business_area_code_Other   -10.600578
8                business_area_code_SUR     2.468903


- Interpret the Coefficients
Each coefficient represents the effect of a unit change in the feature on the log-transformed sales revenue, holding other features constant.

1. value_quantity_log_plus1_transformed (-0.5968): The negative coefficient here suggests that, after accounting for other variables, an increase in the quantity (log-transformed) is associated with a slight decrease in revenue. This could indicate that high sales quantities do not necessarily lead to higher revenue, possibly due to discounts on bulk purchases.
2. value_cost_log_plus1_transformed (1.6040): This positive coefficient indicates that as the cost (log-transformed) increases, revenue increases as well. This is expected, as higher-cost items are often sold at higher prices.
3. value_price_adjustment (-180.59): This significant negative coefficient implies that higher price adjustments (discounts) reduce sales revenue. Discounts appear to have a negative impact on revenue, which might suggest that they reduce profit without sufficiently boosting sales volume.
4. day_gap (0.0342): The small positive coefficient suggests that a longer gap between order and invoice is slightly associated with higher revenue. This could indicate that higher-value items or large orders take longer to process.
- Regional Codes:
1. business_area_code_LMP (4.2870) and business_area_code_SUR (2.4689): These positive coefficients suggest that sales revenue is higher in these regions, indicating potentially profitable markets.
2. business_area_code_DLT (-1.6112) and business_area_code_Other (-10.6006): These negative coefficients indicate that revenue might be lower in these regions, suggesting that these areas are less profitable for LuminaTech.

- Convert Boolean Columns to Numeric: convert them to integers (0 and 1), which are compatible with VIF calculations.

In [24]:
X = X.astype(float)

- Check for Multicollinearity Using Variance Inflation Factor (VIF)

In [26]:
# Ensure all columns in X are numeric
X = X.apply(pd.to_numeric, errors='coerce')

# Calculate VIF for each feature without an explicit loop
vif_data = pd.DataFrame({
    "Feature": X.columns,
    "VIF": X.apply(lambda i: variance_inflation_factor(X.values, X.columns.get_loc(i.name)))
})

print(vif_data)

                                                                   Feature  \
value_quantity_log_plus1_transformed  value_quantity_log_plus1_transformed   
value_cost_log_plus1_transformed          value_cost_log_plus1_transformed   
value_price_adjustment                              value_price_adjustment   
day_gap                                                            day_gap   
business_area_code_DLT                              business_area_code_DLT   
business_area_code_LMP                              business_area_code_LMP   
business_area_code_OTH                              business_area_code_OTH   
business_area_code_Other                          business_area_code_Other   
business_area_code_SUR                              business_area_code_SUR   

                                           VIF  
value_quantity_log_plus1_transformed  1.062628  
value_cost_log_plus1_transformed      1.056339  
value_price_adjustment                1.025621  
day_gap                

- VIF (Variance Inflation Factor)
The VIF values are used to assess multicollinearity among features. A VIF value greater than 5 typically indicates a high level of multicollinearity.

- VIFs for all features are close to 1, with none exceeding 2. This suggests that multicollinearity is not a concern in this model, meaning the independent variables are not strongly correlated with each other. This is desirable as it indicates that each feature contributes uniquely to the model’s predictive power.

- Perform Cross-Validation

In [29]:
# Perform cross-validation with 5 folds and evaluate the R-squared
cv_scores = cross_val_score(model, X, y, cv=5, scoring='r2')
print("Cross-Validation R-squared Scores:", cv_scores)
print("Mean R-squared Score:", cv_scores.mean())

Cross-Validation R-squared Scores: [0.84264484 0.86136497 0.64104513 0.36437966 0.09098071]
Mean R-squared Score: 0.560083060777745


- Cross-Validation
1. Cross-Validation R-squared Scores: The cross-validation scores are relatively high in some folds (0.84 and 0.86), while they drop considerably in others (0.36 and 0.01). This variability suggests that the model’s performance is inconsistent across different subsets of the data, which could indicate that the model may not generalize well to new data.
2. Mean R-squared Score (0.5601): The average cross-validation R-squared score is higher than the R-squared from the initial model, indicating that, on average, the model explains around 56% of the variance in revenue across different subsets. However, due to the variability in cross-validation scores, there might still be room to improve the model’s stability.

Summary and Recommendations
- Insights:
The negative impact of value_price_adjustment on revenue suggests that discounts might be eroding revenue without driving sufficient additional volume. LuminaTech may want to review its discount strategy.
The positive coefficients for business_area_code_LMP and business_area_code_SUR indicate that these regions are strong revenue drivers, suggesting potential areas to focus marketing efforts.
value_quantity_log_plus1_transformed has a negative relationship with revenue, which could imply that high-quantity sales are not necessarily profitable, possibly due to bulk discounts or lower-priced items.

Recommendations:
- Reassess the Discount Strategy: Since value_price_adjustment negatively impacts revenue, LuminaTech should consider targeted discount strategies, potentially reducing discounts in low-margin products or regions.
- Focus on High-Performing Regions: The positive impact of regions like LMP and SUR on revenue suggests that LuminaTech could focus resources or marketing efforts in these areas to maximize revenue.
- Explore Additional Variables: Given the relatively low R-squared, additional factors influencing sales revenue should be considered for analysis, such as customer demographics, seasonal trends, or more granular product features.

#### 2. Question 2: What Factors Influence Profit Margin Across Product Categories and Customer Segments?


Objective
- This question aims to identify factors that impact profit margin across different product types and customer segments. Understanding the factors that lead to higher profit margins enables LuminaTech to focus on high-margin products, optimize discount strategies, and target profitable customer segments. Insights from this analysis can aid in pricing, product prioritization, and regional strategies to maximize profitability.

- Define X with Log-Transformed Columns

In [35]:
# Define X without a loop, using transformed versions of value_cost and value_quantity
X = data[['value_cost_log_plus1_transformed', 'value_quantity_log_plus1_transformed', 'value_price_adjustment', 
                  'light_source_LED', 'light_source_Traditional', 
                  'business_area_code_DLT', 'business_area_code_LMP', 
                  'business_area_code_OTH', 'business_area_code_Other', 
                  'business_area_code_SUR']]

y = data['Profit_Margin']  # Replace with the calculated column if needed

- Split Data into Training and Testing Sets

In [37]:
# Perform an 80-20 split for training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

- Build and Fit the Multiple Regression Model

In [39]:
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

- Evaluate the Model’s Performance

In [41]:
y_pred = model.predict(X_test)
r2 = r2_score(y_test, y_pred)
rmse = mean_squared_error(y_test, y_pred, squared=False)

print(f"R-squared: {r2}")
print(f"RMSE: {rmse}")

R-squared: 0.2106005711708553
RMSE: 17.415020842023182


R-squared & RMSE
- R-squared (0.2106): This indicates that approximately 21.06% of the variance in Profit_Margin is explained by the selected features. This is a relatively low R-squared value, which implies that other factors, not included in the model, may also play a significant role in determining profit margins.
- RMSE (17.42): The Root Mean Squared Error suggests that there is an average error of 17.42 in predicting the profit margin, which might be substantial depending on the scale of profit margin values. This level of error suggests that the model's predictions may not be highly accurate.

- Interpret the Coefficients

In [44]:
coefficients = pd.DataFrame({"Feature": X.columns, "Coefficient": model.coef_})
print(coefficients)

                                Feature  Coefficient
0      value_cost_log_plus1_transformed     0.000606
1  value_quantity_log_plus1_transformed    -0.001771
2                value_price_adjustment    50.280524
3                      light_source_LED    -8.881002
4              light_source_Traditional    -3.731200
5                business_area_code_DLT     5.323921
6                business_area_code_LMP    11.004612
7                business_area_code_OTH    -0.302864
8              business_area_code_Other     2.014051
9                business_area_code_SUR     3.742928


Interpret the Coefficients

1. value_cost_log_plus1_transformed (0.0006): This small positive coefficient suggests a minimal positive impact on profit margin with increased cost. However, the effect size is extremely small, indicating it’s not a strong driver.
2. value_quantity_log_plus1_transformed (-0.0018): This negative coefficient suggests that higher sales quantities are associated with a slight decrease in profit margin, possibly due to discounts on bulk purchases that reduce profitability.
3. value_price_adjustment (50.28): This positive and substantial coefficient indicates that price adjustments (discounts) are actually increasing the profit margin. This could imply that discounts are effectively encouraging profitable sales or that they’re applied to high-margin products.
4. light_source_LED (-8.88) and light_source_Traditional (-3.73): Both product types have negative coefficients, suggesting they may slightly reduce profit margin. However, LED products have a stronger negative impact, which might indicate higher costs or competitive pricing challenges in the LED segment.
- Regional Codes:
1. business_area_code_LMP (11.00) and business_area_code_SUR (3.74): These positive coefficients indicate that these regions are associated with higher profit margins, identifying them as more profitable customer segments.
2. business_area_code_DLT (5.32) and business_area_code_Other (2.01): Both regions also show positive impacts on profit margin, although smaller than LMP and SUR. These regions may still be worth focusing on, though they are less impactful.

- Convert Boolean Columns to Numeric: convert them to integers (0 and 1), which are compatible with VIF calculations.

In [47]:
X = X.astype(float)

- Check for Multicollinearity Using Variance Inflation Factor (VIF)

In [49]:
# Ensure all columns in X are numeric
X = X.apply(pd.to_numeric, errors='coerce')

# Calculate VIF for each feature without an explicit loop
vif_data = pd.DataFrame({
    "Feature": X.columns,
    "VIF": X.apply(lambda i: variance_inflation_factor(X.values, X.columns.get_loc(i.name)))
})

print(vif_data)

                                                                   Feature  \
value_cost_log_plus1_transformed          value_cost_log_plus1_transformed   
value_quantity_log_plus1_transformed  value_quantity_log_plus1_transformed   
value_price_adjustment                              value_price_adjustment   
light_source_LED                                          light_source_LED   
light_source_Traditional                          light_source_Traditional   
business_area_code_DLT                              business_area_code_DLT   
business_area_code_LMP                              business_area_code_LMP   
business_area_code_OTH                              business_area_code_OTH   
business_area_code_Other                          business_area_code_Other   
business_area_code_SUR                              business_area_code_SUR   

                                           VIF  
value_cost_log_plus1_transformed      1.056442  
value_quantity_log_plus1_transformed  1.062

VIF (Variance Inflation Factor)

- VIF for light_source_Traditional (5.03) and business_area_code_LMP (3.80)**: Although these values are higher than others, they are still below critical thresholds, suggesting multicollinearity is not severe in this model.
All other VIFs are close to 1, meaning that multicollinearity is not a significant issue for most variables. Each variable appears to contribute uniquely to the model.

- Perform Cross-Validation

In [52]:
# Perform cross-validation with 5 folds and evaluate the R-squared
cv_scores = cross_val_score(model, X, y, cv=5, scoring='r2')
print("Cross-Validation R-squared Scores:", cv_scores)
print("Mean R-squared Score:", cv_scores.mean())

Cross-Validation R-squared Scores: [0.25716496 0.16010633 0.22576452 0.20225858 0.18277347]
Mean R-squared Score: 0.20561356775196


Cross-Validation
1. Cross-Validation R-squared Scores: The cross-validation scores vary, but they are all relatively low, ranging between 0.16 and 0.26, indicating that the model’s performance is somewhat inconsistent across different folds.
2. Mean R-squared Score (0.2056): The average cross-validation R-squared score aligns closely with the initial R-squared, confirming that the model explains only a small portion of the variability in profit margin. This suggests that the selected features have a limited effect on profit margin, and additional variables may be needed for better predictive power.

Summary and Recommendations

1. Insights:
Discounting appears to increase profit margin, as suggested by the positive coefficient for value_price_adjustment. This could indicate that discounts are well-targeted or that they are applied in ways that still maintain profitability.
Certain regions (LMP and SUR) have a strong positive influence on profit margin, suggesting that LuminaTech could focus resources on these areas.
Higher quantities sold slightly reduce profit margins, potentially due to bulk discounts or reduced prices for large orders.
LED products negatively impact profit margins more than traditional products, which might warrant a review of pricing or cost management strategies for the LED product line.
2. Recommendations:
- Targeted Discounts: Since discounts seem to support profit margin, LuminaTech could further optimize discount strategies to balance revenue and profit.
- Regional Focus: Focus efforts on the most profitable regions, like LMP and SUR, where margins are higher.
Reevaluate LED Product Pricing: Since LED products are associated with lower profit margins, LuminaTech could explore cost-reduction strategies, potentially by negotiating with suppliers or adjusting pricing.